File: src/Robot_urdf/Robot_urdf/vel_driver.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
from std_msgs.msg import Float64
import math
from rcl_interfaces.msg import ParameterDescriptor, ParameterType

class DoubleAckermanController(Node):
    """
    ROS2 node to control a double Ackerman steering vehicle.
    Subscribes to cmd_vel and publishes steering angles and wheel velocities.
    """
    def __init__(self):
        super().__init__('double_ackerman_controller')
        
        # Declare and get parameters with descriptions
        self.declare_and_get_parameters()
        
                # Log the loaded parameters
        self.get_logger().info(f"Loaded parameters: wheel_base={self.wheel_base}, track_width={self.track_width}, " + 
                              f"wheel_radius={self.wheel_radius}, max_steering_angle={self.max_steering_angle}, " +
                              f"max_wheel_speed={self.max_wheel_speed}")
        
    def declare_and_get_parameters(self):
        """
        Declare and get parameters with appropriate descriptions
        """
        # Create parameter descriptors
        wheel_base_desc = ParameterDescriptor(
            type=ParameterType.PARAMETER_DOUBLE,
            description='Distance between front and rear wheels in meters')
            
        track_width_desc = ParameterDescriptor(
            type=ParameterType.PARAMETER_DOUBLE,
            description='Distance between left and right wheels in meters')
            
        wheel_radius_desc = ParameterDescriptor(
            type=ParameterType.PARAMETER_DOUBLE,
            description='Wheel radius in meters')
            
        max_steering_angle_desc = ParameterDescriptor(
            type=ParameterType.PARAMETER_DOUBLE,
            description='Maximum steering angle in radians')
            
        max_wheel_speed_desc = ParameterDescriptor(
            type=ParameterType.PARAMETER_DOUBLE,
            description='Maximum wheel angular velocity in rad/s')
            
        # Declare parameters with default values and descriptors
        self.declare_parameter('wheel_base', 0.8, wheel_base_desc)
        self.declare_parameter('track_width', 0.6, track_width_desc)
        self.declare_parameter('wheel_radius', 0.15, wheel_radius_desc)
        self.declare_parameter('max_steering_angle', 0.6, max_steering_angle_desc)
        self.declare_parameter('max_wheel_speed', 10.0, max_wheel_speed_desc)
        
        # Get parameters
        self.wheel_base = self.get_parameter('wheel_base').value
        self.track_width = self.get_parameter('track_width').value
        self.wheel_radius = self.get_parameter('wheel_radius').value
        self.max_steering_angle = self.get_parameter('max_steering_angle').value
        self.max_wheel_speed = self.get_parameter('max_wheel_speed').value
        
    def cmd_vel_callback(self, msg):
        
        # Subscribe to cmd_vel topic
        self.cmd_vel_sub = self.create_subscription(
            Twist,
            '/cmd_vel',
            self.cmd_vel_callback,
            10)
        
        # Create publishers for steering angles
        self.fl_steer_pub = self.create_publisher(
            Float64, 
            '/fwids/front_left_steer_rad/command', 
            10)
        self.fr_steer_pub = self.create_publisher(
            Float64, 
            '/fwids/front_right_steer_rad/command', 
            10)
        self.rl_steer_pub = self.create_publisher(
            Float64, 
            '/fwids/rear_left_steer_rad/command', 
            10)
        self.rr_steer_pub = self.create_publisher(
            Float64, 
            '/fwids/rear_right_steer_rad/command', 
            10)
        
        # Create publishers for wheel velocities
        self.fl_vel_pub = self.create_publisher(
            Float64, 
            '/fwids/front_left_rotor_radpersec/command', 
            10)
        self.fr_vel_pub = self.create_publisher(
            Float64, 
            '/fwids/front_right_rotor_radpersec/command', 
            10)
        self.rl_vel_pub = self.create_publisher(
            Float64, 
            '/fwids/rear_left_rotor_radpersec/command', 
            10)
        self.rr_vel_pub = self.create_publisher(
            Float64, 
            '/fwids/rear_right_rotor_radpersec/command', 
            10)
        
        self.get_logger().info('Double Ackerman Controller initialized')
        
    def cmd_vel_callback(self, msg):
        """
        Process incoming cmd_vel messages and convert to steering angles and wheel velocities
        
        Args:
            msg: Twist message with linear and angular velocity components
        """
        linear_x = msg.linear.x
        angular_z = msg.angular.z
        
        # If vehicle is not moving, don't steer
        if abs(linear_x) < 0.001 and abs(angular_z) < 0.001:
            self.set_zero_velocity_and_steering()
            return
            
        # Calculate steering angles and wheel velocities
        steering_angles = self.calculate_steering_angles(linear_x, angular_z)
        wheel_velocities = self.calculate_wheel_velocities(linear_x, angular_z, steering_angles)
        
        # Publish steering angles
        self.publish_steering_angles(steering_angles)
        
        # Publish wheel velocities
        self.publish_wheel_velocities(wheel_velocities)
        
    def calculate_steering_angles(self, linear_x, angular_z):
        """
        Calculate the steering angles for each wheel based on the Ackerman geometry
        
        Args:
            linear_x: Linear velocity in x direction
            angular_z: Angular velocity around z axis
            
        Returns:
            Dictionary with steering angles for each wheel
        """
        steering_angles = {
            'front_left': 0.0,
            'front_right': 0.0,
            'rear_left': 0.0,
            'rear_right': 0.0
        }
        
        # If we're not turning, wheels should be straight
        if abs(angular_z) < 0.001:
            return steering_angles
            
        # If we're turning in place, use a simple symmetric model
        if abs(linear_x) < 0.001:
            turn_direction = 1.0 if angular_z > 0 else -1.0
            # Front wheels turn one way, rear wheels turn the opposite way for spot turning
            steering_angles['front_left'] = turn_direction * self.max_steering_angle
            steering_angles['front_right'] = turn_direction * self.max_steering_angle
            steering_angles['rear_left'] = -turn_direction * self.max_steering_angle
            steering_angles['rear_right'] = -turn_direction * self.max_steering_angle
            return steering_angles
        
        # Calculate turning radius based on linear and angular velocity
        # R = v / ω
        # If angular_z is very small, use a very large radius (almost straight)
        if abs(angular_z) < 0.001:
            turning_radius = float('inf')
        else:
            turning_radius = abs(linear_x) / abs(angular_z)
        
        # Determine turn direction
        turn_direction = 1.0 if angular_z > 0 else -1.0
        if linear_x < 0:
            turn_direction *= -1  # Reverse the steering when going backward
        
        # Calculate steering angles using Ackerman geometry
        # For the front wheels
        alpha_front = math.atan(self.wheel_base / (turning_radius - (turn_direction * self.track_width / 2)))
        alpha_front_inner = math.atan(self.wheel_base / (turning_radius - (turn_direction * self.track_width)))
        
        # For the rear wheels (in double Ackerman, rear wheels also steer)
        alpha_rear = math.atan(self.wheel_base / (turning_radius + (turn_direction * self.track_width / 2)))
        alpha_rear_inner = math.atan(self.wheel_base / (turning_radius + (turn_direction * self.track_width)))
        
        # Assign angles based on turn direction
        if turn_direction > 0:  # Turning left
            steering_angles['front_left'] = alpha_front_inner
            steering_angles['front_right'] = alpha_front
            steering_angles['rear_left'] = -alpha_rear_inner
            steering_angles['rear_right'] = -alpha_rear
        else:  # Turning right
            steering_angles['front_left'] = -alpha_front
            steering_angles['front_right'] = -alpha_front_inner
            steering_angles['rear_left'] = alpha_rear
            steering_angles['rear_right'] = alpha_rear_inner
        
        # Clamp steering angles to maximum
        for wheel in steering_angles:
            steering_angles[wheel] = max(-self.max_steering_angle, min(self.max_steering_angle, steering_angles[wheel]))
        
        return steering_angles
    
    def calculate_wheel_velocities(self, linear_x, angular_z, steering_angles):
        """
        Calculate the required wheel velocities based on desired motion
        
        Args:
            linear_x: Linear velocity in x direction
            angular_z: Angular velocity around z axis
            steering_angles: Dictionary with steering angles for each wheel
            
        Returns:
            Dictionary with wheel velocities in rad/s
        """
        wheel_velocities = {
            'front_left': 0.0,
            'front_right': 0.0,
            'rear_left': 0.0,
            'rear_right': 0.0
        }
        
        # If not moving, return zero velocities
        if abs(linear_x) < 0.001 and abs(angular_z) < 0.001:
            return wheel_velocities
        
        # Base velocity from linear component (v/r)
        base_wheel_velocity = linear_x / self.wheel_radius
        
        # Calculate instantaneous center of rotation (ICR)
        if abs(angular_z) < 0.001:
            # Moving straight
            for wheel in wheel_velocities:
                wheel_velocities[wheel] = base_wheel_velocity
        else:
            # Turning - adjust each wheel's velocity based on its distance from the ICR
            # Ensure turning_radius is never zero to avoid division by zero
            if abs(angular_z) < 0.001:
                turning_radius = float('inf')  # Effectively moving straight
            else:
                turning_radius = abs(linear_x) / abs(angular_z)
                
            # Add a safety check to prevent division by zero
            if turning_radius < 0.001:  # If radius is too small (almost rotating in place)
                turning_radius = 0.001  # Set a minimum turning radius
                
            turn_direction = 1.0 if angular_z > 0 else -1.0
            
            # Calculate the distance from each wheel to the ICR
            # This is a simplified model - for a more accurate model, the actual distance 
            # would depend on the steering angle of each wheel
            wheel_velocities['front_left'] = base_wheel_velocity * (1.0 + turn_direction * self.track_width / (2.0 * turning_radius))
            wheel_velocities['front_right'] = base_wheel_velocity * (1.0 - turn_direction * self.track_width / (2.0 * turning_radius))
            wheel_velocities['rear_left'] = base_wheel_velocity * (1.0 + turn_direction * self.track_width / (2.0 * turning_radius))
            wheel_velocities['rear_right'] = base_wheel_velocity * (1.0 - turn_direction * self.track_width / (2.0 * turning_radius))
        
        # Adjust for steering angles - wheels that are steered more should move slower
        for wheel in wheel_velocities:
            angle = steering_angles[wheel]
            # Apply cosine correction factor - a steered wheel needs to rotate faster
            # to maintain the same forward velocity
            if abs(angle) > 0.001:
                wheel_velocities[wheel] = wheel_velocities[wheel] / math.cos(angle)
        
        # Clamp wheel velocities to maximum
        for wheel in wheel_velocities:
            wheel_velocities[wheel] = max(-self.max_wheel_speed, min(self.max_wheel_speed, wheel_velocities[wheel]))
        
        return wheel_velocities
    
    def publish_steering_angles(self, steering_angles):
        """
        Publish calculated steering angles to their respective topics
        
        Args:
            steering_angles: Dictionary with steering angles for each wheel
        """
        fl_msg, fr_msg, rl_msg, rr_msg = Float64(), Float64(), Float64(), Float64()
        
        fl_msg.data = float(steering_angles['front_left'])
        fr_msg.data = float(steering_angles['front_right'])
        rl_msg.data = float(steering_angles['rear_left'])
        rr_msg.data = float(steering_angles['rear_right'])
        
        self.fl_steer_pub.publish(fl_msg)
        self.fr_steer_pub.publish(fr_msg)
        self.rl_steer_pub.publish(rl_msg)
        self.rr_steer_pub.publish(rr_msg)
        
    def publish_wheel_velocities(self, wheel_velocities):
        """
        Publish calculated wheel velocities to their respective topics
        
        Args:
            wheel_velocities: Dictionary with wheel velocities for each wheel
        """
        fl_msg, fr_msg, rl_msg, rr_msg = Float64(), Float64(), Float64(), Float64()
        
        fl_msg.data = float(wheel_velocities['front_left'])
        fr_msg.data = float(wheel_velocities['front_right'])
        rl_msg.data = float(wheel_velocities['rear_left'])
        rr_msg.data = float(wheel_velocities['rear_right'])
        
        self.fl_vel_pub.publish(fl_msg)
        self.fr_vel_pub.publish(fr_msg)
        self.rl_vel_pub.publish(rl_msg)
        self.rr_vel_pub.publish(rr_msg)
    
    def set_zero_velocity_and_steering(self):
        """
        Set all wheel velocities and steering angles to zero
        """
        zero_msg = Float64()
        zero_msg.data = 0.0
        
        # Publish zero steering
        self.fl_steer_pub.publish(zero_msg)
        self.fr_steer_pub.publish(zero_msg)
        self.rl_steer_pub.publish(zero_msg)
        self.rr_steer_pub.publish(zero_msg)
        
        # Publish zero velocity
        self.fl_vel_pub.publish(zero_msg)
        self.fr_vel_pub.publish(zero_msg)
        self.rl_vel_pub.publish(zero_msg)
        self.rr_vel_pub.publish(zero_msg)

def main(args=None):
    rclpy.init(args=args)
    node = DoubleAckermanController()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        # Zero out all commands before shutting down
        node.set_zero_velocity_and_steering()
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/vel_driver/src/vel_driver.cpp
#include "vel_driver/vel_driver.hpp"
#include <cmath>
#include <algorithm>

namespace gazebo
{

// constructor
VelDriver::VelDriver()
    : Node("vel_driver")
{
    // load parameters 

    //// Double Ackerman vehicle parameters
    this->declare_parameter<float>("wheel_base", 2.39);
    this->declare_parameter<float>("track_width", 1.34);
    this->declare_parameter<float>("wheel_radius", 0.4);
    this->declare_parameter<float>("max_steering_angle", 0.7);
    this->declare_parameter<float>("velocity_threshold", 1e-2);

    this->get_parameter("wheel_base", wheel_base);
    this->get_parameter("track_width", track_width);
    this->get_parameter("wheel_radius", wheel_radius);
    this->get_parameter("max_steering_angle", max_steering_angle);
    this->get_parameter("velocity_threshold", velocity_threshold);

    // Parameter validation to prevent division by zero and other issues
    if (std::abs(wheel_radius) < 1e-9) {
        RCLCPP_ERROR(this->get_logger(), "Invalid wheel_radius parameter (%f). Must be non-zero. Setting to default value 0.4.", wheel_radius);
        wheel_radius = 0.4;
    }
    if (wheel_base <= 0.0) {
        RCLCPP_ERROR(this->get_logger(), "Invalid wheel_base parameter (%f). Must be positive. Setting to default value 2.39.", wheel_base);
        wheel_base = 2.39;
    }
    if (track_width <= 0.0) {
        RCLCPP_ERROR(this->get_logger(), "Invalid track_width parameter (%f). Must be positive. Setting to default value 1.34.", track_width);
        track_width = 1.34;
    }
    if (velocity_threshold <= 0.0) {
        RCLCPP_WARN(this->get_logger(), "Invalid velocity_threshold parameter (%f). Must be positive. Setting to default value 1e-2.", velocity_threshold);
        velocity_threshold = 1e-2;
    }
    
    RCLCPP_INFO(this->get_logger(), "Double Ackerman Parameters: wheelbase=%.2fm, track_width=%.2fm, wheel_radius=%.2fm, max_steer=%.2frad", 
                wheel_base, track_width, wheel_radius, max_steering_angle);

    //// subscribing topic names
    this->declare_parameter<std::string>("control_cmd_vel_topic", "/cmd_vel");
    std::string control_cmd_vel_topic;
    this->get_parameter("control_cmd_vel_topic", control_cmd_vel_topic);

    //// publishing topic names
    this->declare_parameter<std::string>("front_left_steer_cmd_topic", "/fwids/front_left_steer_rad/command");
    this->declare_parameter<std::string>("front_right_steer_cmd_topic", "/fwids/front_right_steer_rad/command");
    this->declare_parameter<std::string>("rear_left_steer_cmd_topic", "/fwids/rear_left_steer_rad/command");
    this->declare_parameter<std::string>("rear_right_steer_cmd_topic", "/fwids/rear_right_steer_rad/command");
    this->declare_parameter<std::string>("front_left_rotor_cmd_topic", "/fwids/front_left_rotor_radpersec/command");
    this->declare_parameter<std::string>("front_right_rotor_cmd_topic", "/fwids/front_right_rotor_radpersec/command");
    this->declare_parameter<std::string>("rear_left_rotor_cmd_topic", "/fwids/rear_left_rotor_radpersec/command");
    this->declare_parameter<std::string>("rear_right_rotor_cmd_topic", "/fwids/rear_right_rotor_radpersec/command");

    std::string front_left_steer_cmd_topic, front_right_steer_cmd_topic, rear_left_steer_cmd_topic, rear_right_steer_cmd_topic;
    std::string front_left_rotor_cmd_topic, front_right_rotor_cmd_topic, rear_left_rotor_cmd_topic, rear_right_rotor_cmd_topic;
    
    this->get_parameter("front_left_steer_cmd_topic", front_left_steer_cmd_topic);
    this->get_parameter("front_right_steer_cmd_topic", front_right_steer_cmd_topic);
    this->get_parameter("rear_left_steer_cmd_topic", rear_left_steer_cmd_topic);
    this->get_parameter("rear_right_steer_cmd_topic", rear_right_steer_cmd_topic);
    this->get_parameter("front_left_rotor_cmd_topic", front_left_rotor_cmd_topic);
    this->get_parameter("front_right_rotor_cmd_topic", front_right_rotor_cmd_topic);
    this->get_parameter("rear_left_rotor_cmd_topic", rear_left_rotor_cmd_topic);
    this->get_parameter("rear_right_rotor_cmd_topic", rear_right_rotor_cmd_topic);

    // initialize subscribers
    sub_cmd_vel_ = this->create_subscription<geometry_msgs::msg::Twist>(
        control_cmd_vel_topic,1, std::bind(&VelDriver::cmdVelCallback, this, std::placeholders::_1));
    // initialize publishers
    pub_cmd_front_left_steer = this->create_publisher<std_msgs::msg::Float64>(front_left_steer_cmd_topic,1);
    pub_cmd_front_right_steer = this->create_publisher<std_msgs::msg::Float64>(front_right_steer_cmd_topic,1);
    pub_cmd_rear_left_steer = this->create_publisher<std_msgs::msg::Float64>(rear_left_steer_cmd_topic,1);
    pub_cmd_rear_right_steer = this->create_publisher<std_msgs::msg::Float64>(rear_right_steer_cmd_topic,1);
    pub_cmd_front_left_rotor = this->create_publisher<std_msgs::msg::Float64>(front_left_rotor_cmd_topic,1);
    pub_cmd_front_right_rotor = this->create_publisher<std_msgs::msg::Float64>(front_right_rotor_cmd_topic,1);
    pub_cmd_rear_left_rotor = this->create_publisher<std_msgs::msg::Float64>(rear_left_rotor_cmd_topic,1);
    pub_cmd_rear_right_rotor = this->create_publisher<std_msgs::msg::Float64>(rear_right_rotor_cmd_topic,1);
}

// destructor
VelDriver::~VelDriver()
{
    // No Contents
}

// Helper function to calculate Ackerman steering angles for left and right wheels
void VelDriver::calculateAckermannAngles(float turning_radius, bool is_front, 
                                         float &left_angle, float &right_angle)
{
    // For straight motion (infinite turning radius)
    if (std::isinf(turning_radius) || std::abs(turning_radius) > 1e6) {
        left_angle = 0.0;
        right_angle = 0.0;
        return;
    }
    
    // Ackerman geometry: inner and outer wheel angles differ
    // The wheel on the inside of the turn has a larger steering angle
    float half_track = track_width / 2.0;
    float abs_radius = std::abs(turning_radius);
    
    // Safety check: ensure we don't divide by very small numbers
    // The inner wheel radius must be greater than half the track width
    if (abs_radius <= half_track) {
        RCLCPP_WARN(this->get_logger(), 
            "Turning radius (%.2f m) too small compared to track width (%.2f m). Setting angles to max.",
            abs_radius, track_width);
        // Set to maximum steering angle
        left_angle = (turning_radius > 0) ? max_steering_angle : -max_steering_angle;
        right_angle = left_angle * 0.7; // Outer wheel angle is smaller
        if (!is_front) {
            left_angle = -left_angle;
            right_angle = -right_angle;
        }
        return;
    }
    
    if (turning_radius > 0) {
        // Turning left (counter-clockwise)
        // Inner wheel (left) has larger angle, outer wheel (right) has smaller angle
        left_angle = std::atan(wheel_base / (abs_radius - half_track));   // inner wheel
        right_angle = std::atan(wheel_base / (abs_radius + half_track));  // outer wheel
    } else {
        // Turning right (clockwise)
        // Inner wheel (right) has larger angle, outer wheel (left) has smaller angle
        right_angle = std::atan(wheel_base / (abs_radius - half_track));      // inner wheel
        left_angle = std::atan(wheel_base / (abs_radius + half_track));       // outer wheel
        // Make angles negative for right turn
        left_angle = -left_angle;
        right_angle = -right_angle;
    }
    
    // For rear axle in double Ackerman, angles are opposite sign
    // This creates the characteristic "crab steering" behavior
    if (!is_front) {
        left_angle = -left_angle;
        right_angle = -right_angle;
    }
    
    // Clamp to maximum steering angle for safety
    left_angle = std::max(-max_steering_angle, std::min(max_steering_angle, left_angle));
    right_angle = std::max(-max_steering_angle, std::min(max_steering_angle, right_angle));
}

// /cmd_vel topic callback
void VelDriver::cmdVelCallback(const geometry_msgs::msg::Twist::SharedPtr msg)
{
    // if msg contains NaN, return without publishing commands
    if (std::isnan(msg->linear.x) || std::isnan(msg->linear.y) || std::isnan(msg->angular.z))
    {
        RCLCPP_WARN(this->get_logger(), "Received NaN in Twist Command. Skip this message.");
        return;
    }

    // parse received twist message
    float v_x = msg->linear.x;   // forward velocity
    float v_y = msg->linear.y;   // lateral velocity (should be ~0 for Ackerman)
    float omega = msg->angular.z; // angular velocity around the z-axis

    // [for debug] announce received Twist message
    RCLCPP_DEBUG(this->get_logger(), "Received Twist: vx=%.2f, vy=%.2f, omega=%.2f", v_x, v_y, omega);

    // Check if robot is commanded to stop
    bool is_stopping = (std::abs(v_x) < velocity_threshold && 
                       std::abs(v_y) < velocity_threshold && 
                       std::abs(omega) < velocity_threshold);

    if (is_stopping) {
        // Stop all wheels
        cmd_front_left_rotor.data = 0.0;
        cmd_front_right_rotor.data = 0.0;
        cmd_rear_left_rotor.data = 0.0;
        cmd_rear_right_rotor.data = 0.0;
        
        // Keep last steering angles (don't change them when stopped)
        RCLCPP_DEBUG(this->get_logger(), "Robot stopping - all wheels set to zero velocity");
    } else {
        // Double Ackerman steering kinematics
        // Calculate turning radius from angular velocity
        float turning_radius;
        
        if (std::abs(omega) < 1e-6) {
            // Straight line motion
            turning_radius = std::numeric_limits<float>::infinity();
        } else {
            // For double Ackerman, the instantaneous center of rotation (ICR) 
            // is located at the intersection of the wheel axes extensions
            // Turning radius = linear_velocity / angular_velocity
            // But we need to ensure it's kinematically feasible
            turning_radius = v_x / omega;
            
            // Calculate minimum turning radius based on maximum steering angle
            // For double Ackerman: R_min = wheelbase / tan(max_steering_angle)
            // Using the center of the vehicle as reference
            float min_turning_radius = wheel_base / std::tan(max_steering_angle);
            
            // Add safety margin (half track width) to account for wheel positions
            float safe_min_radius = min_turning_radius + track_width / 2.0;
            
            // Limit the turning radius to kinematically feasible values
            if (std::abs(turning_radius) < safe_min_radius) {
                // Requested turn is too tight, limit to minimum feasible radius
                turning_radius = (turning_radius > 0) ? safe_min_radius : -safe_min_radius;
                RCLCPP_DEBUG(this->get_logger(), 
                    "Requested turning radius too tight, limited to %.2f m (min: %.2f m)", 
                    turning_radius, safe_min_radius);
            }
        }
        
        // Calculate Ackerman steering angles for front and rear axles
        float front_left_angle, front_right_angle;
        float rear_left_angle, rear_right_angle;
        
        calculateAckermannAngles(turning_radius, true, front_left_angle, front_right_angle);
        calculateAckermannAngles(turning_radius, false, rear_left_angle, rear_right_angle);
        
        // Update steering angle commands
        cmd_front_left_steer.data = front_left_angle;
        cmd_front_right_steer.data = front_right_angle;
        cmd_rear_left_steer.data = rear_left_angle;
        cmd_rear_right_steer.data = rear_right_angle;
        
        // Calculate wheel velocities
        // For double Ackerman, all wheels should rotate at velocities that match the turning geometry
        
        if (std::isinf(turning_radius)) {
            // Straight line - all wheels same speed
            float wheel_speed_base = v_x / wheel_radius;
            cmd_front_left_rotor.data = wheel_speed_base;
            cmd_front_right_rotor.data = wheel_speed_base;
            cmd_rear_left_rotor.data = wheel_speed_base;
            cmd_rear_right_rotor.data = wheel_speed_base;
        } else {
            // Different speeds for inner and outer wheels during turning
            // Each wheel travels along a circular arc with radius dependent on its position
            float half_track = track_width / 2.0;
            float half_wheelbase = wheel_base / 2.0;
            
            // Calculate the actual distance from the instantaneous center of rotation (ICR)
            // to each wheel contact point
            // ICR is on the line connecting the midpoints of front and rear axles
            float abs_turning_radius = std::abs(turning_radius);
            
            // Distance from ICR to each wheel (using Pythagorean theorem)
            // For left turn (turning_radius > 0), ICR is on the left side
            // For right turn (turning_radius < 0), ICR is on the right side
            float r_fl, r_fr, r_rl, r_rr;
            
            if (turning_radius > 0) {
                // Left turn: left wheels are inner, right wheels are outer
                r_fl = std::sqrt(std::pow(abs_turning_radius - half_track, 2) + std::pow(half_wheelbase, 2));
                r_fr = std::sqrt(std::pow(abs_turning_radius + half_track, 2) + std::pow(half_wheelbase, 2));
                r_rl = std::sqrt(std::pow(abs_turning_radius - half_track, 2) + std::pow(half_wheelbase, 2));
                r_rr = std::sqrt(std::pow(abs_turning_radius + half_track, 2) + std::pow(half_wheelbase, 2));
            } else {
                // Right turn: right wheels are inner, left wheels are outer
                r_fl = std::sqrt(std::pow(abs_turning_radius + half_track, 2) + std::pow(half_wheelbase, 2));
                r_fr = std::sqrt(std::pow(abs_turning_radius - half_track, 2) + std::pow(half_wheelbase, 2));
                r_rl = std::sqrt(std::pow(abs_turning_radius + half_track, 2) + std::pow(half_wheelbase, 2));
                r_rr = std::sqrt(std::pow(abs_turning_radius - half_track, 2) + std::pow(half_wheelbase, 2));
            }
            
            // Calculate angular velocity about the turning center
            // omega is already given, but we recalculate from the actual turning radius
            // to ensure consistency with kinematic constraints
            float actual_omega = v_x / turning_radius;  // rad/s
            float omega_abs = std::abs(actual_omega);
            
            // Linear velocity at each wheel = radius * angular_velocity
            // Then convert to wheel angular velocity = linear_velocity / wheel_radius
            cmd_front_left_rotor.data = (r_fl * omega_abs) / wheel_radius;
            cmd_front_right_rotor.data = (r_fr * omega_abs) / wheel_radius;
            cmd_rear_left_rotor.data = (r_rl * omega_abs) / wheel_radius;
            cmd_rear_right_rotor.data = (r_rr * omega_abs) / wheel_radius;
            
            // Preserve direction of motion (forward/backward)
            if (v_x < 0) {
                cmd_front_left_rotor.data = -cmd_front_left_rotor.data;
                cmd_front_right_rotor.data = -cmd_front_right_rotor.data;
                cmd_rear_left_rotor.data = -cmd_rear_left_rotor.data;
                cmd_rear_right_rotor.data = -cmd_rear_right_rotor.data;
            }
        }
        
        RCLCPP_DEBUG(this->get_logger(), "Turning radius: %.2f, Steer: FL=%.3f FR=%.3f RL=%.3f RR=%.3f", 
                     turning_radius, front_left_angle, front_right_angle, rear_left_angle, rear_right_angle);
    }

    // Publish commands
    pub_cmd_front_left_steer->publish(cmd_front_left_steer);
    pub_cmd_front_right_steer->publish(cmd_front_right_steer);
    pub_cmd_rear_left_steer->publish(cmd_rear_left_steer);
    pub_cmd_rear_right_steer->publish(cmd_rear_right_steer);

    pub_cmd_front_left_rotor->publish(cmd_front_left_rotor);
    pub_cmd_front_right_rotor->publish(cmd_front_right_rotor);
    pub_cmd_rear_left_rotor->publish(cmd_rear_left_rotor);
    pub_cmd_rear_right_rotor->publish(cmd_rear_right_rotor);
}
}

